packages/piro/src/Piro_NOXSolver_Def.hpp

```diff
@@ -222,459 +244,462 @@ void Piro::NOXSolver<Scalar>::evalModelImpl(
-
-          auto xbar = createMembers(dgdx->range(), num_cols);
-          {
-            const auto timer = Teuchos::rcp(new Teuchos::TimeMonitor(*Teuchos::TimeMonitor::getNewTimer("Piro::NOXSolver::evalModelImpl::calc xbar")));
-            xbar->assign(0);
-            Thyra::solve(
-                *jacobian,
-                Thyra::NOTRANS,
-                *dgdx,
-                xbar.ptr(),
-                Teuchos::ptr(&solve_criteria));
-
-            Thyra::scale(-1.0, xbar.ptr());
+
+      if(Teuchos::nonnull(prec))
+        Thyra::initializePreconditionedOp<double>(*lows_factory,
+            Thyra::transpose<double>(lop),
+            Thyra::unspecifiedPrec<double>(::Thyra::transpose<double>(prec->getUnspecifiedPrecOp())),
+            jacobian.ptr());
+      else
+        Thyra::initializeOp<double>(*lows_factory,
+            Thyra::transpose<double>(lop),
+            jacobian.ptr());
+
+
+      for (int j=0; j<num_g; j++) {
+
+        // See if there are any forward sensitivities we need to do
+        // that aren't handled by the operator
+        do_sens = false;
+        for (int i=0; i<num_p; i++) {
+          if (!outArgs.supports(Thyra::ModelEvaluatorBase::OUT_ARG_DgDp, j, i).none()) {
+            if (outArgs.get_DgDp(j,i).getMultiVector() != Teuchos::null)
+              do_sens = true;
           }
+        }
+        if (!do_sens)
+          continue;
 
-          // (3) Calculate dg/dp^T = df/dp^T*xbar + dg/dp^T
-          for (int i=0; i<num_p; i++) {
-            std::string paramName = "Parameter " + std::to_string(i);
-            if(appParams->isSublist("Analysis")){
-              auto analysisParams = appParams->sublist("Analysis");
-              if (analysisParams.isSublist("ROL")) {
-                std::ostringstream ss; ss << "Parameter Vector Index " << i;
-                auto rolParams = analysisParams.sublist("ROL");
-                const int p_ind = rolParams.get(ss.str(), i);
-                const auto paramNames = *this->getModel().get_p_names(p_ind);
-                paramName = paramNames[0];
-              }
+        if (!modelOutArgs.supports(Thyra::ModelEvaluatorBase::OUT_ARG_DgDx, j).none()) {
+          TEUCHOS_TEST_FOR_EXCEPTION(
+              modelOutArgs.get_DgDx(j).getLinearOp()!=Teuchos::null,
+              std::logic_error,
+              std::endl << "Piro::NOXSolver::evalModel():  " <<
+              "Can\'t use dg/dx operator " << j << " with non-operator " <<
+              "adjoint sensitivities." << std::endl);
+          auto dgdx  = modelOutArgs.get_DgDx(j).getMultiVector();
+          if (dgdx != Teuchos::null) {
+            int num_cols = dgdx->domain()->dim();
+
+            // (2) Calculate xbar multivector from -(J^{-T}*dg/dx)
+
+            auto xbar = createMembers(dgdx->range(), num_cols);
+            {
+              const auto timer = Teuchos::rcp(new Teuchos::TimeMonitor(*Teuchos::TimeMonitor::getNewTimer("Piro::NOXSolver::evalModelImpl::calc xbar")));
```

Comment:
Please don't handle TimeMonitor by pointer if you can avoid it (which you easily can here).  TimeMonitor is a scope guard.  Its constructor starts the timer, and its destructor stops it.  You already create a new scope here.  Why not use the scope guard as a scope guard, directly?

Reply:
Will do.